projects/cli-files/go-cat/cmd/root.go

```diff
@@ -0,0 +1,108 @@
+package cmd
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"strings"
+)
+
+func Execute() {
+	// define a flag of type boolean with the specific properties (name, value, usage)
+	numberFlag := flag.Bool("n", false, "number all output lines")
+
+	// parse the command line flags
+	flag.Parse()
+
+	// if there is no argument given then error
+	if len(os.Args) < 2 {
+		fmt.Fprintf(os.Stderr, "no filename provided\n")
+		os.Exit(2)
+	}
+
+	// create a variable to store the file name
+	var filename string
+
+	if *numberFlag {
+		// if the -n flag is provided the filename will be the third value in the os.Args slice
+		filename = os.Args[2]
+	} else {
+		// if no flag is provided the filename will be the second value in the os.Args slice
+		filename = os.Args[1]
+	}
+
+	// try to get the file information
+	fileInfo, err := os.Stat(filename)
+
+	// if we cannot get the file information then error
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "could not read the file/directory information: %s\n", filename)
+		os.Exit(2)
+	}
+
+	// if the file is a directory then warn the user
+	if fileInfo.IsDir() {
+		fmt.Fprintf(os.Stderr, "%s : Is a directory\n", filename)
+		os.Exit(1)
+	}
+
+	// try to read the file
+	fileContents, err := os.ReadFile(filename)
+
+	// if we cannot read the file then error
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to read the file: %v\n", err)
+		os.Exit(2)
+	}
+
+	// convert the file contents to a string
+	fileContentsAsString := string(fileContents)
+
+	// if the -n flag was provided
+	if *numberFlag {
+		// split the string into lines using the newline delimiter
+		lines := strings.Split(fileContentsAsString, "\n")
+		// loop over the lines printing them with a line number prefix
+		for index, line := range lines {
+			fmt.Fprintf(os.Stdout, "%d  %s\n", index+1, line)
```

Comment:
What does `cat` do if you give it a file which _doesn't_ end in a newline?

What does your program do?

Are those things the same?

You can test this by running the below (also, work out what the below is actually doing, it's a combination of quite interesting things!):

```console
% cat -n <(echo -n hello) 
     1	hello%                                                                   
% ./go-cat -n <(echo -n hello)
1  hello
```